### ISSUE TITLE
Can I use YOLO's pose detection to detect keypoints in charts?

### ISSUE THREAD

**Issue by starbilibili** [ultralytics/ultralytics#18145]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello, I want to use YOLO's pose detection model to detect keypoints in charts, but the training results are very poor, mainly due to the following issues:
(1) There is a problem with the weight settings of the loss function. With the default configuration, pose: 12 and kobj: 1, it is found that the pose loss cannot be reduced at all during training. Later, I tried setting pose: 0 and kobj: 10, and during training, the kobj loss decreased normally, but the model performance is still very poor.
![image](https://github.com/user-attachments/assets/c14147e8-0845-4bc5-8625-0deee5796be4)

(2) There is an issue with the dataset format. Due to the significant variation in the number of keypoints across different samples, with some having fewer than 10 keypoints and others having up to 200, the model requires a consistent number of keypoints. As a result, I had to pad with 0,0,0 to represent invalid keypoints, but I am uncertain whether this practice might negatively impact training, given the large number of invalid keypoints and the small number of valid keypoints.

I want to get the result like this:
![image](https://github.com/user-attachments/assets/ff0e3310-bfca-40c0-9a08-b4031f2b10ed)
I hope you can help me determine if YOLO is capable of handling this task. If it is, what should be my subsequent adjustment strategy?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @starbilibili, thank you for your interest in Ultralytics 🚀! We recommend visiting the [Docs](https://docs.ultralytics.com) where you'll find detailed resources for [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples, which may address some of your questions.

Regarding your inquiry, if this is a ❓ Question or related to a custom training scenario, it would be helpful if you could share more detailed information about your setup including your configuration files, training commands, and any logs generated during training. Also, we encourage you to verify that you are adhering to our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

For the 🐛 issues you described—such as the loss function weights not converging or the dataset format challenges—it would also be immensely beneficial to provide a [minimum reproducible example (MRE)](https://docs.ultralytics.com/help/minimum_reproducible_example/). The MRE can help us understand your dataset structure, preprocessing logic, annotations, and model configuration, so we can assist more effectively.

💡 In the meantime, here are a few steps for you to try:

## Upgrade

Make sure you are using the latest version of `ultralytics` to capture all recent updates and optimizations. Upgrade the package in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) using:

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in multiple user-friendly and verified environments. If you haven't tried these yet, they might make your workflow easier:

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM: See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI: See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**: See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Community Support

Additionally, feel free to join the Ultralytics community if you haven't already! Get real-time help or in-depth discussions here:
- 🎧 [Discord](https://discord.com/invite/ultralytics) for live support and chat
- [Discourse](https://community.ultralytics.com) for thorough Q&A threads
- [Subreddit](https://reddit.com/r/Ultralytics) for knowledge sharing in the community

## Status

Finally, make sure you're checking the latest repository CI status for potential issues.  
<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

An Ultralytics engineer will review your issue and assist further shortly. Thank you for exploring the potential of YOLO models for pose detection—we’re excited to see how this progresses! 😊

**Comment by Y-T-G** [comment]
Maybe you should treat each bar as an object instead of the whole image. Otherwise, the keypoints wouldn't be consistent. The keypoints require consistency. The same keypoint can't be assigned to different visual feature for each image. Just like if a keypoint is for left hand, it should always be assigned to left hand.

The loss doesn't decrease because your labeling is inconsistent so the model can't learn anything.

**Comment by starbilibili** [comment]
> Maybe you should treat each bar as an object instead of the whole image. Otherwise, the keypoints wouldn't be consistent. The keypoints require consistency. The same keypoint can't be assigned to different visual feature for each image. Just like if a keypoint is for left hand, it should always be assigned to left hand.
> 
> The loss doesn't decrease because your labeling is inconsistent so the model can't learn anything.

well, very useful ideas, I will try it through experiments. Thank you for your reply!

**Comment by glenn-jocher** [comment]
@starbilibili you're welcome! I'm glad the suggestions were helpful. Feel free to share your results or reach out if you have further questions. Best of luck with your experiments!

**Comment by starbilibili** [comment]
> @starbilibili you're welcome! I'm glad the suggestions were helpful. Feel free to share your results or reach out if you have further questions. Best of luck with your experiments!

Experiments have shown that when dealing with bar charts, assigning a separate detection box to each bar can significantly enhance the training effectiveness. However, when it comes to pie charts and line charts, it is necessary to treat the entire chart as a whole and assign only one detection box. But I am still facing some issues; there is a significant disparity in the number of keypoints across different types of charts: bars have only 2 keypoints, while pie and line charts are limited to 16 and 32 keypoints, respectively. When I combine the datasets for training, the model still struggles to converge, and I am further experimenting to explore solutions.

**Comment by glenn-jocher** [comment]
Thank you for the update! It seems the significant disparity in keypoint counts across chart types might be causing inconsistencies during training. One potential approach is to train separate models for each chart type to ensure consistency in keypoint structure and distribution. Alternatively, normalizing the keypoint data (e.g., padding and masking invalid keypoints) and setting task-specific loss weights could help improve convergence. Feel free to share additional details if you'd like further guidance!

**Comment by starbilibili** [comment]
The keypoint detection logic of YOLO seems to involve detecting bounding boxes first, and then detecting the keypoints of the objects within those bounding boxes. When I was training on line charts, I found that YOLO has difficulty handling line charts with overlapping lines. Additionally, when there are abrupt changes in the slope, the model also fails to predict accurately, tending to detect keypoints with a trend of gentle variation. I will next attempt to adjust the weights of the loss function, but achieving good results seems challenging. Perhaps I should consider using a corner detection method instead.
![f80b84b2d4941dbf3406727dadde4114_d3d3LnNjYi5zZQkxOTMuMTgwLjEyNC45MQ== xls-3-0](https://github.com/user-attachments/assets/95ffcc1c-d79b-466f-984e-6adeae8fa857)

![f74a7c52d461bbbed1a492577051d83f_d3d3LnNlYWZpc2gub3JnCTQ2LjI1NS4xMTUuMjUw-4-3](https://github.com/user-attachments/assets/e89a5f37-dc2f-401e-9036-02b2397afa44)

**Comment by glenn-jocher** [comment]
Thank you for sharing your observations. You are correct that YOLO's keypoint detection relies on bounding boxes as an intermediate step, which can struggle with complex scenarios like overlapping lines or abrupt slope changes in line charts. Adjusting the loss weights might help marginally, but for such cases, specialized corner detection or graph-based methods may indeed be better suited. If you decide to continue with YOLO, refining your dataset and exploring task-specific augmentations or pre-processing might also be beneficial. Let us know if you need further assistance!
